File: 322.零钱兑换.cpp
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */

// @lc code=start
class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (int i = 1; i <= coins.size(); i++)
        {
            for (int j = coins[i - 1]; j <= amount; j++)
            {
                if (dp[j] - 1 > dp[j - coins[i - 1]])
                    dp[j] = 1 + dp[j - coins[i - 1]];
            }
        }
        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};
// @lc code=end

File: JavaScript/322.零钱兑换.js
/*
 * @lc app=leetcode.cn id=322 lang=javascript
 *
 * [322] 零钱兑换
 */

// @lc code=start
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */

// dp[j] = Math.min(dp[j], dp[j - coins[i - 1]] + 1);
var coinChange = function (coins, amount) {
  const dp = new Array(amount + 1).fill(Number.MAX_SAFE_INTEGER);
  dp[0] = 0;

  for (let i = 1; i <= coins.length; i++) {
    for (let j = coins[i - 1]; j <= amount; j++) {
      dp[j] = Math.min(dp[j], dp[j - coins[i - 1]] + 1);
    }
  }

  return dp[amount] === Number.MAX_SAFE_INTEGER ? -1 : dp[amount];
};
// @lc code=end
